src/api/source/mvt.js

```diff
@@ -0,0 +1,175 @@
+        const mvtLayer = tile.layers[Object.keys(tile.layers)[0]];
+
+        const { points, featureGeometries, properties } = this._decodeMVTLayer(mvtLayer, this._metadata, mvt_extent);
+
+
+        const rs = rsys.getRsysFromTile(x, y, z);
+        const dataframeGeometry = this._metadata.geomType == geometryTypes.POINT ? points : featureGeometries;
+        const dataframe = this._generateDataFrame(rs, dataframeGeometry, properties, mvtLayer.length, this._metadata.geomType);
+        return dataframe;
+    }
+
+
+    _decodeMVTLayer(mvtLayer, metadata, mvt_extent) {
+        let points;
+        if (metadata.geomType == geometryTypes.POINT) {
+            points = new Float32Array(mvtLayer.length * 2);
+        }
+        let featureGeometries = [];
+        const decodedProperties = {};
+        for (let i = 0; i < mvtLayer.length; i++) {
+            const f = mvtLayer.feature(i);
+            const geom = f.loadGeometry();
+            const mvtGeomType = f.type;
+            if (metadata.geomType === undefined) {
+                switch (mvtGeomType) {
+                    case 1:
+                        metadata.geomType = geometryTypes.POINT;
+                        break;
+                    case 2:
+                        metadata.geomType = geometryTypes.LINE;
+                        break;
+                    case 3:
+                        metadata.geomType = geometryTypes.POLYGON;
+                        break;
+                    default:
+                        throw new Error('MVT: invalid geometry type');
+                }
+                if (metadata.geomType == geometryTypes.POINT) {
+                    points = new Float32Array(mvtLayer.length * 2);
+                }
+            }
+            if (metadata.geomType == geometryTypes.POINT) {
+                points[2 * i + 0] = 2 * (geom[0][0].x) / mvt_extent - 1.;
+                points[2 * i + 1] = 2 * (1. - (geom[0][0].y) / mvt_extent) - 1.;
+            } else if (metadata.geomType == geometryTypes.POLYGON) {
+                const decodedPolygons = decodePolygons(geom, mvt_extent);
+                featureGeometries.push(decodedPolygons);
+            } else if (metadata.geomType == geometryTypes.LINE) {
+                const decodedLines = decodeLines(geom, mvt_extent);
+                featureGeometries.push(decodedLines);
+            } else {
+                throw new Error(`Unimplemented geometry type: '${metadata.geomType}'`);
+            }
+            Object.keys(f.properties).forEach(propertyName => {
+                const propertyValue = f.properties[propertyName];
+                const decodedPropertyValue = this.decodeProperty(propertyName, propertyValue);
+                if (decodedPropertyValue !== undefined) {
+                    if (decodedProperties[propertyName] === undefined) {
+                        decodedProperties[propertyName] = new Float32Array(mvtLayer.length + 1024);
```

Comment:
Why `1024`?

Reply:
Ups, I should have used the constant `RTT_WIDTH` defined in `renderer.js`. The reason for this addition is that when we use the property Float32Array in `dataframe.js` to upload it to WebGL there is a required padding of the width of the texture, which is equal to `RTT_WIDTH`. Adding `RTT_WIDTH` assures that the padding exists. I'll document this reason on the code.